src/peer_select.cc

```diff
@@ -83,6 +77,122 @@ class PeerSelectionDumper
+/// alternative would add thousands of concurrent events to the Squid event queue
+/// instead.
+class PeerSelectorTimeoutProcessor
+{
+public:
+    static void NoteWaitOver(void *raw);
+
+    /// calls back all ready PeerSelectors and continues to wait for others
+    void noteWaitOver();
+
+    /// adds a PeerSelector to the waiting list
+    void enqueue(PeerSelector *selector, const PeerSelectAbsoluteTime timeout);
+
+    /// removes a PeerSelector from the waiting list
+    void dequeue(PeerSelector *selector);
+
+private:
+    void wait(const PeerSelectAbsoluteTime);
+    bool waiting() { return waitEnd_ > 0; }
+
+    PeerSelectAbsoluteTime waitStart_ = 0; ///< the time of the last wait() call
+    PeerSelectAbsoluteTime waitEnd_ = 0; ///< expected NoteWaitOver() call time (or zero)
+    dlink_list peerSelectors; /// a list of postponed PeerSelectors
+};
+
+PeerSelectorTimeoutProcessor ThePeerSelectorTimeoutProcessor;
+
+static void
+HandlePingTimeout(PeerSelector *selector)
+{
+    selector->handlePingTimeout();
+}
+
+void
+PeerSelectorTimeoutProcessor::NoteWaitOver(void *raw)
+{
+    assert(raw);
+    static_cast<PeerSelectorTimeoutProcessor*>(raw)->noteWaitOver();
+}
+
+void
+PeerSelectorTimeoutProcessor::noteWaitOver()
+{
+    assert(waitEnd_ >= waitStart_);
+    const PeerSelectAbsoluteTime lastTimeout = waitEnd_ - waitStart_;
+    waitEnd_ = 0;
+    auto link = peerSelectors.head;
+    while (link) {
+        auto selector = reinterpret_cast<PeerSelector *>(link->data);
+        link = link->next;
+        if (selector->peerWaiting.ready()) {
+            selector->peerWaiting.stop(&peerSelectors);
+            AsyncCall::Pointer callback = asyncCall(44, 4, "HandlePingTimeout", cbdataDialer(HandlePingTimeout, selector));
+            callback->codeContext = selector->peerWaiting.codeContext;
+            ScheduleCallHere(callback);
+        }
+    }
+    if (peerSelectors.head)
+        wait(lastTimeout);
+}
+
+void
+PeerSelectorWait::start(PeerSelector *selector, const PeerSelectAbsoluteTime timeout, dlink_list *list)
+{
+    waitEnd = current_dtime + timeout;
+    codeContext = CodeContext::Current();
+    dlinkAdd(selector, &peerSelectorNode, list);
+}
+
+void
+PeerSelectorWait::stop(dlink_list *list)
+{
+    waitEnd = 0;
+    dlinkDelete(&peerSelectorNode, list);
+}
+
+bool
+PeerSelectorWait::ready() const
+{
+    assert(waitEnd > 0);
+    return waitEnd <= current_dtime;
+}
+
+void
+PeerSelectorTimeoutProcessor::wait(const PeerSelectAbsoluteTime interval)
+{
+    assert(!waiting());
+    waitStart_ = current_dtime;
+    waitEnd_ = waitStart_ + interval;
+    eventAdd("PeerSelectorTimeoutProcessor::NoteWaitOver", &PeerSelectorTimeoutProcessor::NoteWaitOver, this, interval, 0, false);
+}
+
+void
+PeerSelectorTimeoutProcessor::enqueue(PeerSelector *selector, const PeerSelectAbsoluteTime timeout)
+{
+    selector->peerWaiting.start(selector, timeout, &peerSelectors);
+    if (!waiting())
+        wait(timeout);
```

Comment:
Order: This "start waiting on the first enqueue() call" logic works only if all subsequent enqueue() calls will use the same (or higher) timeout values. After looking at how neighborsUdpPing() computes the timeout, I doubt that is true. The actual timeout may depend on the number of factors beyond configuration, such as the number of expected replies (that may be different for different peers and/or peer selectors?).

FWIW, IIRC, HappyConnOpener code was specifically designed to avoid such problems. It uses constant timeouts. However, even that code needs to be more careful than this PR. See HappyOrderEnforcer::startedWaiting() for a similar but more complex `!waiting` condition.

Reply:
> This "start waiting on the first enqueue() call" logic works only if all subsequent enqueue() calls will use the same (or higher) timeout values.

Agreed. As you saw, marked the problem of selecting a correct timeout before ThePeerSelectorTimeoutProcessor.enqueue() invocation (as a TODO) - this should be an XXX.

> The actual timeout may depend on the number of factors beyond configuration

Agreed. Though neighborsUdpPing() dynamically calculates a timeout, there is a minimum timeout value, which is constantly Config.Timeout.icp_query (if specified),  or Config.Timeout.icp_query_min. We can use this constant minimum for the event schedule. We will need to cancel the event and re-schedule it, after these parameters changing, due to reconfiguration.